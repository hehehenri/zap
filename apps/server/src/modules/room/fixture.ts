import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
