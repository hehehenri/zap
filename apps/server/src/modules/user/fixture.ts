import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
