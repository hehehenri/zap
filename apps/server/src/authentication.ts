import jwt from "jsonwebtoken";
import { UserModel, UserDocument } from "./modules/user/UserModel";
import config from "./config";
import { KoaContext } from "./context";

export type Config = {
  secret: string
}

type TokenPayload = {
  id: string
}

export const getToken = (ctx: KoaContext) => {
  const auth = ctx.headers.authorization;
  if (!auth) return null;

  if (auth.startsWith("Bearer ")) {
    return auth.substring(7, auth.length);
  }

  return null
}

export const getAuth = async (token: string | null) => {
  if (!token) return { user: null };

  try {
    const decoded = jwt.verify(token, config.jwt.secret) as TokenPayload;

    const user = await UserModel.findById(decoded.id);
    return { user };
  } catch (_) {
    return { user: null };
  }
}

export const generateToken = (user: UserDocument) => {
  return jwt.sign({ id: user._id }, config.jwt.secret);
}
