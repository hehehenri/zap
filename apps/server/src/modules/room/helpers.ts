import { UserDocument } from "../user/UserModel";
import { RoomDocument } from "./RoomModel";

export const isRoomMember = (room: RoomDocument, user: UserDocument) => {
  return room
    .participants
    .map(u => u._id.toString())
    .includes(user._id.toString());
}
