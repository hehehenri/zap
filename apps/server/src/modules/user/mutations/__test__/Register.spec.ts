import { createUser } from "../../fixture";
import { getAuth } from "../../../../authentication";
import { describeWithDb, testQuery } from "../../../../test/helpers";


const query = ({ username, password }: { username?: string, password?: string }) => ({
  query: `
    mutation Register($username: String! $password: String!) {
      register(input: { username: $username, password: $password}) { token }
    }
  `,
  variables: {
    username: username ?? "username",
    password: password ?? "password",
  }
});

describeWithDb('user/mutations/register', () => {
  it("should create user and return user token and info", async () => {      
    const response = await testQuery({ query: query({}) })
    expect(response.status).toBe(200);

    const data = response.body.data.register;
    const { user: tokenUser } = await getAuth(data.token);

    expect(tokenUser?.username).toBe("username");
  });

  it(
    "should return invalid payload when username already exists",
    async () => {
      const queryData = { username: "username", password: "password" };
      await createUser(queryData);

      const response = await testQuery({ query: query(queryData) });

      expect(response.status).toBe(422);
      expect(response.body.errors).toMatchSnapshot();
    });
  }
);
