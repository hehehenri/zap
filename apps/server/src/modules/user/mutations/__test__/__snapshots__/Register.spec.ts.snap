// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user/mutations/register should return invalid payload when username already exists 1`] = `
[
  {
    "locations": [
      {
        "column": 7,
        "line": 3,
      },
    ],
    "message": "Username already registered",
    "path": [
      "register",
    ],
  },
]
`;
